onnx/ir.h

```diff
@@ -0,0 +1,1048 @@
+    return elem_type_;
+  }
+  Value* setSizes(const std::vector<Dimension>& sizes) {
+    sizes_ = sizes;
+    return this;
+  }
+  const std::vector<Dimension>& sizes() const {
+    return sizes_;
+  }
+  size_t unique() const {
+    return unique_;
+  }
+  bool has_name() const {
+    return has_name_;
+  }
+  std::string uniqueName() const {
+    if(unique_name_.size() > 0)
+      return unique_name_;
+    return std::to_string(unique());
+  }
+  Value* setUniqueName(const std::string & name) {
+    has_name_ = true;
+    unique_name_ = name;
+    return this;
+  }
+  Value* setStage(size_t s) {
+    stage_ = s;
+    return this;
+  }
+  size_t stage() const {
+    return stage_;
+  }
+  Node* node() {
+    return node_;
+  }
+  size_t offset() const {
+    return offset_;
+  }
+  const Node * node() const {
+    return node_;
+  }
+  Graph * owningGraph();
+  const Graph * owningGraph() const;
+  // TODO: make this more const correct
+  const use_list & uses() const {
+    return uses_;
+  }
+
+  // Replaces all uses of this node with 'newValue'.
+  //
+  // Given:   %3 = f(%1, %2)
+  //          %4 = g(%3)
+  //          %5 = h(%3, %3)
+  // Execute: %3.replaceAllUsesWith(%6)
+  // Result:  %3 = f(%1, %2)
+  //          %4 = g(%6)
+  //          %5 = h(%6, %6)
+  void replaceAllUsesWith(Value * newValue);
+
+  Value* copyMetadata(Value * from) {
+    setElemType(from->elemType());
+    setSizes(from->sizes());
+    if (!from->unique_name_.empty()) {
+      setUniqueName(from->uniqueName());
+    }
+    return this;
+  }
+
+};
+
+
+struct Node : public Attributes<Node> {
+  ONNX_DISALLOW_COPY_AND_ASSIGN(Node);
+  friend struct Graph;
+  friend struct Value;
+  friend graph_node_list;
+  friend const_graph_node_list;
+  friend graph_node_list_iterator;
+  friend const_graph_node_list_iterator;
+
+private:
+  // each node but Return/Param
+  // is associated with exactly one place in the node list...
+  // of the graph_
+  // this circular is a doubly-linked list, the Return node is used as the sentinel for the beginning and end of the list
+  // such that the list never has null pointers
+  // next_in_graph[0] is next pointer
+  // next_in_graph[1] is prev pointer
+  // using an array to allow the same iterator class for forward and reverse node lists
+  // This list represents a topological sort
+
+  Node* next_in_graph[2] = { nullptr, nullptr };
+  Node* & next() { return next_in_graph[kNextDirection]; }
+  Node* & prev() { return next_in_graph[kPrevDirection]; }
+  Node* const & next() const { return next_in_graph[kNextDirection]; }
+  Node* const & prev() const { return next_in_graph[kPrevDirection]; }
+
+  const NodeKind kind_;
+  std::vector<Value*> inputs_;
+  std::vector<Value*> outputs_;
+  Graph* graph_;
+  size_t stage_;
+  bool has_name_;
+  std::string name_;
+  bool has_doc_string_;
+  std::string doc_string_;
+
+protected:
+  Node(Graph * graph_, NodeKind kind_); //defined after graph
+
+public:
+  bool has_name() {
```

Comment:
Since we allow the empty string as name of the input/output to resolve the optional parameter list issue, here we have to distinguish empty string and missing string.